Wheel event doesn't get fired on an element that has been reinserted after document.open

rniwa opened the pull request:
#### a39166fd0ab3b6d329a33075ef2ddbfdcd0ca89f
<pre>
Wheel event doesn&apos;t get fired on an element that has been reinserted after document.open
<a href="https://bugs.webkit.org/show_bug.cgi?id=267399">https://bugs.webkit.org/show_bug.cgi?id=267399</a>

Reviewed by Wenson Hsieh.

The bug was caused by Document::removeAllEventListeners falsely assuming that all wheel and touch
event listeners from all nodes associated with this document are removed. This is not the case.
Document::removeAllEventListeners does not remove any event listeners from disconnected nodes.

Fixed the bug by relying on newly added Node::removeAllEventListeners to remove itself from
Document&apos;s maps of wheel and touch event listeners instead of clearing them all in
Document::removeAllEventListeners. Node::removeAllEventListeners won&apos;t be called on any
disconnected nodes and they would remain in the document&apos;s maps.

An alternative approach is for Document to keep track of only connected Nodes which have touch
and wheel event listeners but this will involve adding code to Node::insertedIntoAncestor and
Node::removedFromAncestor to add or remove nodes from Document&apos;s maps. We opt not to adopt this
approach since it incurs runtime cost of O(k) where k is the maximum number of event listeners.

* LayoutTests/fast/events/touch/ios/touch-event-after-document-open-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/touch-event-after-document-open.html: Added.
* LayoutTests/fast/events/wheel/wheel-event-after-document-open-expected.txt: Added.
* LayoutTests/fast/events/wheel/wheel-event-after-document-open.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::removeAllEventListeners):
* Source/WebCore/dom/Node.cpp:
(WebCore::didRemoveEventListenerOfType): Extracted from tryRemoveEventListener.
(WebCore::tryRemoveEventListener): Deleted.
(WebCore::Node::removeEventListener):
(WebCore::Node::removeAllEventListeners): Added.
* Source/WebCore/dom/Node.h:

Canonical link: <a href="https://commits.webkit.org/272960@main">https://commits.webkit.org/272960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a63369224066b29ea5543cfbeb6d4f0813858aa0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12508 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35673 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30588 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14882 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29662 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30077 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9204 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9301 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30094 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37673 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30620 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30415 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35429 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9442 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7385 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33316 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11222 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29753 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10025 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4345 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10228 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on LayoutTests/fast/events/touch/ios/touch-event-after-document-open.html:
```diff
@@ -0,0 +1,48 @@
+
+    // FIXME: This shouldn't be necesasry but somehow it is.
+    if (window.testRunner)
+        testRunner.notifyDone();
```
...interesting! I suppose, because of the `document.write()` or `document.close()`? 🤔

rniwa replied:
I have no idea! I've written other tests with document.open/write but those don't need extra notifyDone call either so it's very mysterious.

cdumez commented on LayoutTests/fast/events/touch/ios/touch-event-after-document-open.html:
```diff
@@ -0,0 +1,48 @@
+
+    finishJSTest();
+
+    // FIXME: This shouldn't be necesasry but somehow it is.
```
Typo: necesasry

rniwa replied:
Oh oops, will fix.

rniwa committed:
Wheel event doesn't get fired on an element that has been reinserted after document.open
https://bugs.webkit.org/show_bug.cgi?id=267399

Reviewed by Wenson Hsieh.

The bug was caused by Document::removeAllEventListeners falsely assuming that all wheel and touch
event listeners from all nodes associated with this document are removed. This is not the case.
Document::removeAllEventListeners does not remove any event listeners from disconnected nodes.

Fixed the bug by relying on newly added Node::removeAllEventListeners to remove itself from
Document's maps of wheel and touch event listeners instead of clearing them all in
Document::removeAllEventListeners. Node::removeAllEventListeners won't be called on any
disconnected nodes and they would remain in the document's maps.

An alternative approach is for Document to keep track of only connected Nodes which have touch
and wheel event listeners but this will involve adding code to Node::insertedIntoAncestor and
Node::removedFromAncestor to add or remove nodes from Document's maps. We opt not to adopt this
approach since it incurs runtime cost of O(k) where k is the maximum number of event listeners.

* LayoutTests/fast/events/touch/ios/touch-event-after-document-open-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/touch-event-after-document-open.html: Added.
* LayoutTests/fast/events/wheel/wheel-event-after-document-open-expected.txt: Added.
* LayoutTests/fast/events/wheel/wheel-event-after-document-open.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::removeAllEventListeners):
* Source/WebCore/dom/Node.cpp:
(WebCore::didRemoveEventListenerOfType): Extracted from tryRemoveEventListener.
(WebCore::tryRemoveEventListener): Deleted.
(WebCore::Node::removeEventListener):
(WebCore::Node::removeAllEventListeners): Added.
* Source/WebCore/dom/Node.h:

Canonical link: https://commits.webkit.org/272960@main
